File: etudes/er/geom.erl
%% @author Young King <yanckin@gmail.com> [http://youngking.org]
%% @doc Functions for calculatings of geometric shapes
%% @copyright 2013 Young King
%% @version 0.1

-module(geom).
-export([area/1]).

%% @doc Caclcuates the area of a shape, given the tuple containing shape
%% and two of the dimensions. 

- spec(area({atom(), number(), number()}) -> number() ).

area({Shape, L, W}) -> area(Shape, L, W).


%% @doc Caclcuates the area of a shape, given the shape
%% and two of the dimensions. 

area(Shape, A, B) when A>=0, B>=0 ->
    Ratio = case Shape of 
        rectangle -> 1.0;
        triangle -> 0.5;
        ellipse -> math:pi()
    end,

    Ratio * A * B.

File: etudes/eli/geom.ex
defmodule Geom do
  @moduledoc """
  Functions for calculating areas of geometric shapes.
  """
  @vsn 0.1

  @doc """
  Calculates the area of rectangle, given the length and width.
  Returns the production of its arguments. The default value for both
  arguments is 1.
  """
  def area({shape, dim1, dim2}) when dim1>=0 and dim2>=0 do
    case shape do
      :rectangle -> dim1 * dim2
      :triangle ->  dim1 * dim2 / 2.0
      :ellipse -> :math.pi * dim1 * dim2
    end
  end

end
